core/src/sumeragi/main_loop.rs

```diff
@@ -517,90 +504,77 @@ fn handle_message(
         (Message::ViewChangeSuggested, _) => {
             trace!("Received view change suggestion.");
         }
-        (Message::BlockSyncUpdate(block_sync_update), _) => {
-            let block_hash = block_sync_update.hash();
-            info!(%addr, %role, hash=%block_hash, "Block sync update received");
-
-            let block = match block_sync_update
-                .clone()
-                .validate_and_extract_block::<false>(
-                    &sumeragi.transaction_validator,
-                    state.wsv.clone(),
-                    state.latest_block_hash,
-                    state.latest_block_height,
-                )
-                .or_else(|_|
-                /* If the block fails validation we must check again using the finaziled wsv.
-                When a soft-fork occurs the consensus-block may be valid on the previous
-                wsv but not the current one. */
-                block_sync_update.validate_and_extract_block::<false>(
-                    &sumeragi.transaction_validator,
-                    state.finalized_wsv.clone(),
-                    state.previous_block_hash,
-                    state.latest_block_height.saturating_sub(1),
-                )) {
+        (Message::BlockSyncUpdate(BlockSyncUpdate { block }), _) => {
+            let block_hash = block.hash();
+            debug!(%addr, %role, %block_hash, "Block sync update received");
+
+            let block = match block.validate(
+                state.latest_block_height,
+                state.latest_block_hash,
+                &state.current_topology,
+                &sumeragi.transaction_validator,
+                state.wsv.clone(),
+            ) {
                 Ok(block) => block,
-                Err(error) => {
-                    error!(%addr, %role, %block_hash, ?error, "Block not valid.");
-                    return;
+                Err((block, error)) => {
+                    let header = block.header();
+
+                    // TODO: This seems incorrect?
+                    if state.latest_block_height == header.height
+                        && state.previous_block_hash == header.previous_block_hash
+                        && state.latest_block_hash != Some(block_hash)
+                        && state.latest_block_view_change_index < header.view_change_index
+                    {
+                        debug!(%addr, %role, %block_hash, ?error, "Block appears to be invalid. Revalidating...");
+
+                        /* If the block fails validation we must check again using the finaziled wsv.
+                        When a soft-fork occurs the consensus-block may be valid on the previous
+                        wsv but not the current one. */
+                        match block.validate(
+                            state.latest_block_height.saturating_sub(1),
+                            state.previous_block_hash,
+                            &state.current_topology,
```

Comment:
I think in case of soft-fork revalidation we should pass not `current_topology`, but `previous_topology`. Suppose that wrong block added some new peers.
I can try to address this in my #3405.

Reply:
I wanted to ask you to specifically review soft fork resolution code

Reply:
Shouldn't we add a test for that case? Maybe in another PR.

Or there was one test failing because of that?

Reply:
@Arjentix i had some tests for soft fork detection in #3405, but i think we still need to cover case with topology.